## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [27]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [29]:
# View the column names in the data
column_names = df.columns
print(column_names)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [30]:
# Use the describe function to gather some basic statistics
description = df.describe()
print(description)
df.head()

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [31]:
# Use this space to do any additional research
# and familiarize yourself with the data.
print(df.info())
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [33]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)
print("Top Three Item Categories by Entry Count:")
print(top_categories) 

Top Three Item Categories by Entry Count:
category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [37]:
# For the category with the most entries, which subcategory had the most entries?
top_category = df['category'].value_counts().idxmax()
top_subcategory = df[df['category'] == top_category]['subcategory'].value_counts().idxmax()
print(f"For the category '{top_category}', the subcategory '{top_subcategory}' had the most entries.")

For the category 'consumables', the subcategory 'bathroom supplies' had the most entries.


In [38]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
print("Top Five Clients by Entry Count:")
print(top_clients)

Top Five Clients by Entry Count:
client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [39]:
# Store the client ids of those top 5 clients in a list.
top_clients = df['client_id'].value_counts().head(5)
top_client_ids = top_clients.index.tolist()
print("Top Five Client IDs:")
print(top_client_ids)

Top Five Client IDs:
[33615, 66037, 46820, 38378, 24741]


In [40]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id = df['client_id'].value_counts().idxmax()
total_units_ordered = df[df['client_id'] == top_client_id]['qty'].sum()
print(f"The client with the most entries (client ID {top_client_id}) ordered a total of {total_units_ordered} units.")


The client with the most entries (client ID 33615) ordered a total of 64313 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [41]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']

In [43]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calculate_shipping_price(row):
    if row['unit_weight'] is not None and row['unit_weight'] > 50:
        return 7 * row['unit_weight']
    else:
        return 10 * row['unit_weight']

df['shipping_price'] = df.apply(calculate_shipping_price, axis=1)


In [44]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['subtotal'] = df['unit_price'] * df['qty']
df['total_price'] = df['subtotal'] + df['shipping_price']

sales_tax_rate = 0.0925
df['sales_tax'] = df['total_price'] * sales_tax_rate

df['total_price_with_tax'] = df['total_price'] + df['sales_tax']

In [45]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
def calculate_line_cost(row):
    return row['unit_price'] * row['qty'] + row['shipping_price']

df['line_cost'] = df.apply(calculate_line_cost, axis=1)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

    subtotal  shipping_price  total_price     sales_tax  total_price_with_tax  \
0  115164.00            75.0    115239.00  10659.607500         125898.607500   
1     523.95            14.9       538.85     49.843625            588.693625   
2     527.28            16.8       544.08     50.327400            594.407400   
3    1056.18            12.3      1068.48     98.834400           1167.314400   
4    3902.00           464.3      4366.30    403.882750           4770.182750   

   line_cost  
0  115239.00  
1     538.85  
2     544.08  
3    1068.48  
4    4366.30  

[5 rows x 24 columns]

In [46]:
# Create a column for the profit of each line using line cost and line price
def calculate_profit(row):
    return row['unit_price'] - row['line_cost']
df['profit'] = df.apply(calculate_profit, axis=1)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number   subtotal  \
0          17        2023  ...        7.50  105           1  115164.00   
1          20        2023  ...        1.49   21           0     523.95   
2           4        2023  ...        1.68   39           6     527.28   
3          17        2023  ...        1.23   29           3    1056.18   
4          21        2023  ...       46.43   20           1    3902.00   

   shipping_price  total_price     sales_tax  total_price_with_tax  line_cost  \
0            75.0    115239.00  10659.607500         125898.607500  115239.00   
1            14.9       538.85     49.843625            588.693625     538.85   
2            16.8       544.08     50.327400            594.407400     544.08   
3            12.3      1068.48     98.834400           1167.314400    1068.48   
4           464.3      4366.30    403.882750           4770.182750    4366.30   

      profit  
0 -114142.20  
1    -513.90  
2    -530.56  
3   -1032.06  
4   -4171.20  

[5 rows x 25 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [48]:
# Check your work using the totals above
top_5_clients = df['client_id'].value_counts().idxmax()
total_units_ordered = df[df['client_id'] == top_client_id]['qty'].sum()

print(f"The client with the most entries (client ID {top_client_id}) ordered a total of {total_units_ordered} units.")

df['subtotal'] = df['unit_price'] * df['qty']

sales_tax_rate = 0.0925
df['sales_tax'] = df['total_price'] * sales_tax_rate
df['total_price_with_tax'] = df['total_price'] + df['sales_tax']

def calculate_line_cost(row):
    return row['unit_price'] * row['qty'] + row['shipping_price']

df['line_cost'] = df.apply(calculate_line_cost, axis=1)

def calculate_profit(row):
    return row['unit_price'] - row['line_cost']

df['profit'] = df.apply(calculate_profit, axis=1)

The client with the most entries (client ID 33615) ordered a total of 64313 units.


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [49]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
for id in top_5_clients:
    price = df.loc[df['client_id'] == 'unit_price'].sum()
    print(f"{id}: ${price:,.2f}")



TypeError: 'numpy.int64' object is not iterable

In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_5 = df[df['client_id'].isin(top_5_clients)]
summary = top_5.groupby('client_id')[['qty', 'shipping_price', 'unit_price', 'line_cost', 'profit']].sum().reset_index()
summary['client_id'] = pd.Categorical(summary['client_id'], categories=top_5_clients, ordered=True)
summary = summary.sort_values('client_id').reset_index(drop=True)

summary

client_id     qty  shipping_price   line_price    line_cost  line_profit
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31

In [18]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 
summary.columns = ['Client ID', 'Units', 'Shipping (millions)', 'Total Revenue (millions)', 'Total Cost (millions)', 'Total Profit (millions)' ]

def currency_format_millions(x):
    return x/1000000

for column in summary.columns:
    if 'millions' in column:
        summary[column] =  summary[column].apply(currency_format_millions)
    else:
        continue



Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      38378   73667             3.429455                 12.906551   
4      24741  239862             5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [19]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
summary = summary.sort_values(by='Total Profit (millions)', ascending=False)
summary 

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995